src/server/handlers.rs

```diff
@@ -0,0 +1,347 @@
+use async_nats::{Client, Message};
+use serde_json::json;
+use tracing::{debug, error, instrument, trace};
+
+use crate::{model::internal::StoredManifest, storage::Store};
+
+use super::{
+    parser::parse_manifest, DeleteModelRequest, DeleteModelResponse, DeleteResult, GetModelRequest,
+    ModelSummary, StatusType, VersionInfo,
+};
+
+pub(crate) struct Handler<S> {
+    pub(crate) store: S,
+    pub(crate) client: Client,
+}
+
+impl<S: Store + Send + Sync> Handler<S> {
+    #[instrument(level = "debug", skip(self, msg))]
+    pub async fn put_model(&self, msg: Message, lattice_id: &str, name: &str) {
+        trace!("Parsing incoming manifest");
+        let manifest = match parse_manifest(msg.payload.into(), msg.headers.as_ref()) {
+            Ok(m) => m,
+            Err(e) => {
+                self.send_error(msg.reply, format!("Unable to parse manifest: {e:?}"))
+                    .await;
+                return;
+            }
+        };
+        if manifest.metadata.name != name {
+            self.send_error(
+                msg.reply,
+                "Manifest name doesn't match name from topic".to_string(),
+            )
+            .await;
+            return;
+        }
+
+        trace!(
+            ?manifest,
+            "Manifest is valid. Fetching current manifests from store"
+        );
+
+        let mut current_manifests: StoredManifest = match self.store.get(lattice_id, name).await {
+            Ok(d) => d.unwrap_or_default(),
+            Err(e) => {
+                error!(error = %e, "Unable to fetch data from store");
+                self.send_error(msg.reply, "Internal storage error".to_string())
+                    .await;
+                return;
+            }
+        };
+
+        let current_version = manifest.version().to_owned();
```

Comment:
Naming: `current_version` made me think we'd be reading from `current_manifests`, not the incoming model